cpp/src/arrow/python/builtin_convert.cc

```diff
@@ -111,10 +122,22 @@ class ScalarVisitor {
         ss << type->ToString();
         return Status::Invalid(ss.str());
       }
+    } else if (internal::PyDecimal_Check(obj)) {
+      // Don't infer anything if we encounter a Decimal('nan')
+      if (!internal::PyDecimal_ISNAN(obj)) {
+        int32_t precision;
+        int32_t scale;
+        RETURN_NOT_OK(internal::InferDecimalPrecisionAndScale(obj, &precision, &scale));
+        max_decimal_precision_ = std::max(max_decimal_precision_, precision);
+        if (std::abs(max_decimal_scale_) < std::abs(scale)) {
```

Comment:
Might be nice to add a comment motivating the algorithm here (why compare the absolute values but then memorize the original value)?

Reply:
Actually, reconsidering this based on your comment this really should be just the max scale. Negative scale should contribute to precision only if it would increase precision. The goal here is to "cast the widest net", ie the max precision and max scale. Negative scale complicates things a tiny bit. I'll add some commentary.